src/java.base/macosx/classes/sun/nio/fs/BsdFileAttributeViews.java

```diff
@@ -49,28 +50,94 @@ private static void setTimes(UnixPath path, FileTime lastModifiedTime,
         // permission check
         path.checkWrite();
 
-        int commonattr = 0;
-        long modValue = 0L;
-        if (lastModifiedTime != null) {
-            modValue = lastModifiedTime.to(TimeUnit.NANOSECONDS);
-            commonattr |= UnixConstants.ATTR_CMN_MODTIME;
-        }
-        long accValue = 0L;
-        if (lastAccessTime != null) {
-            accValue = lastAccessTime.to(TimeUnit.NANOSECONDS);
-            commonattr |= UnixConstants.ATTR_CMN_ACCTIME;
+        boolean useLutimes = false;
+        try {
+            useLutimes = !followLinks &&
+                UnixFileAttributes.get(path, false).isSymbolicLink();
+        } catch (UnixException x) {
+            x.rethrowAsIOException(path);
         }
-        long createValue = 0L;
-        if (createTime != null) {
-            createValue = createTime.to(TimeUnit.NANOSECONDS);
-            commonattr |= UnixConstants.ATTR_CMN_CRTIME;
+
+        int fd = -1;
+        if (!useLutimes) {
+            try {
+                fd = path.openForAttributeAccess(followLinks);
+            } catch (UnixException x) {
+                x.rethrowAsIOException(path);
+            }
         }
 
         try {
-            setattrlist(path, commonattr, modValue, accValue, createValue,
-                        followLinks ?  0 : UnixConstants.FSOPT_NOFOLLOW);
-        } catch (UnixException x) {
-            x.rethrowAsIOException(path);
+            // not all volumes support setattrlist(2), so set the last
+            // modified and last access times using futimens(2)
+            if (lastModifiedTime != null || lastAccessTime != null) {
+                // if not changing both attributes then need existing attributes
+                if (lastModifiedTime == null || lastAccessTime == null) {
+                    try {
+                        UnixFileAttributes attrs = UnixFileAttributes.get(fd);
+                        if (lastModifiedTime == null)
+                            lastModifiedTime = attrs.lastModifiedTime();
+                        if (lastAccessTime == null)
+                            lastAccessTime = attrs.lastAccessTime();
+                    } catch (UnixException x) {
+                        x.rethrowAsIOException(path);
+                    }
+                }
+
+                // update times
+                TimeUnit timeUnit = useLutimes ?
+                    TimeUnit.MICROSECONDS : TimeUnit.NANOSECONDS;
+                long modValue = lastModifiedTime.to(timeUnit);
+                long accessValue= lastAccessTime.to(timeUnit);
+
+                boolean retry = false;
+                try {
+                    if (useLutimes)
+                        lutimes(path, accessValue, modValue);
+                    else
+                        futimens(fd, accessValue, modValue);
+                } catch (UnixException x) {
+                    // if futimens fails with EINVAL and one/both of the times is
```

Comment:
The comment should probably say "lutimes/futimens".

Reply:
Comment modified in fc589d209bd60a63eb7935ced2be776e12ca223e.